sql: Convert roachpb.Error back to error

kkaneda opened the pull request:
This change partially reverts the previous change that introduced `roachpb.Error` in the sql package. This initial change is conservative, and I'll send follow-up PRs for more changes.

One concrete example that we cannot use error is when an error is used inside `Txn`. For example, `planner.exec` needs to return `roachpb.Error` as the method is called inside `db.Txn()` in `LeaseStore.Acquire()`.

Some other changes:
- Rename `err` to `pErr` at several places where `roachpb.Error` is actually passed.
- Change `makeIndexJoin()` to not return an error since it always return `nil`.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4126)

<!-- Reviewable:end -->

kkaneda committed:
sql: Convert roachpb.Error back to error

This change partially reverts the previous change that introduced
roachpb.Error in the sql package.

This initial change is conservative, and I'll send follow-up PRs for
more changes.

One concrete example that we cannot use error is when an error is used
inside Txn. For example, planner.exec needs to return roachpb.Error as
the method is called inside db.Txn() in LeaseStore.Acquire().

Some other changes:
- Rename err to pErr at several places where roachpb.Error is actually passed.
- Change makeIndexJoin() to no return an error since it always return nil.

tamird commented:
LGTM. I sure hope you have a plan!

---

Reviewed 22 of 22 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 633 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-K9hRxdTY_u6JBzaY9ju:-K9hRxdTY_u6JBzaY9jv:-280913026) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e3b6ab19b5bc4edd3dc4a7966214fd64a731b7a/sql/executor.go#L633)):*
while you're here, i think this should be `util.Errorf`

---

*[sql/limit.go, line 50 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-K9hSIFcx0CC2iU1rkMz:-K9hSIFcx0CC2iU1rkN-:394227909) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e3b6ab19b5bc4edd3dc4a7966214fd64a731b7a/sql/limit.go#L50)):*
util.Errorf

---

*[sql/table.go, line 463 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-K9hSZqxihBtLYMdX2PA:-K9hSZqyJZoxNAUrGh-h:394227909) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e3b6ab19b5bc4edd3dc4a7966214fd64a731b7a/sql/table.go#L463)):*
util.Errorf

---

*[sql/table.go, line 803 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-K9hSddZbX6TJ9o3-jnN:-K9hSddZbX6TJ9o3-jnO:-663836646) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e3b6ab19b5bc4edd3dc4a7966214fd64a731b7a/sql/table.go#L803)):*
util.Errorf?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-:-K9hSiRhbwkptZKljYTf:771740861)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-:-K9i4jJPrxO0PmgJWGl6:676834233)*
<!-- Sent from Reviewable.io -->

kkaneda commented:
Thanks for the reviews! 

Looking into test failures now. It seems the initial range split didn't complete quickly enough. 

```
   ...
W0205 13:35:37.381580 425 kv/dist_sender.go:624  failed to invoke AdminSplit [/Table/12/0,/Min): storage/replica_command.go:1338: split at key /Table/12 failed: sql/descriptor.go:147: "rangelog" is not a table
   ...
I0205 13:35:47.484280 425 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
	testserver.go:59: Could not start server: server/testserver.go:245: had 2 ranges at startup, expected 3
	leaktest.go:110: prior leak detected, leaktest disabled
      
```

Will update once test failures are fixed. Will also continue to fix (e.g., `convertBatchError` make return an `error`) and investigate why we cannot use `roachpb.Error` in certain places. 

---

Review status: 17 of 22 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 633 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-K9hRxdTY_u6JBzaY9ju:-K9jKM5xyVF4WYltFuA6:1900274870) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e3b6ab19b5bc4edd3dc4a7966214fd64a731b7a/sql/executor.go#L633)):*
Hmm, it seems we need to use `fmt.Errorf` here. pg_wire_test.go failed when I changed this to `util.Errorf`.

https://github.com/cockroachdb/cockroach/blob/master/sql/pgwire_test.go#L171

---

*[sql/limit.go, line 50 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-K9hSIFcx0CC2iU1rkMz:-K9jKLgS8gyh1bB-iqS_:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e3b6ab19b5bc4edd3dc4a7966214fd64a731b7a/sql/limit.go#L50)):*
Done.

---

*[sql/table.go, line 463 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-K9hSZqxihBtLYMdX2PA:-K9jKLB7mCtd2J76O2KZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e3b6ab19b5bc4edd3dc4a7966214fd64a731b7a/sql/table.go#L463)):*
Done.

---

*[sql/table.go, line 803 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-K9hSddZbX6TJ9o3-jnN:-K9jK-LPvyq2gwpEuMtm:271004300) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e3b6ab19b5bc4edd3dc4a7966214fd64a731b7a/sql/table.go#L803)):*
it seems this needs to be `fmt.Errorf`. pgwire_test.go failed here as well..

https://github.com/cockroachdb/cockroach/blob/master/sql/pgwire_test.go#L180

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-:-K9jKN2ce0WVjYHY-fw-:-1734371147)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 633 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-K9hRxdTY_u6JBzaY9ju:-K9lu9yiM3fSXU1M8Cmg:-733505645) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e3b6ab19b5bc4edd3dc4a7966214fd64a731b7a/sql/executor.go#L633)):*
Ah, I see. I think the solution is to have a separate clause for DValArg and have it return "could not infer ...", and then the default can be `util.Errorf`. @mjibson?

---

*[sql/table.go, line 803 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-K9hSddZbX6TJ9o3-jnN:-K9luPckIomfNxBxQPKG:-2017399771) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e3b6ab19b5bc4edd3dc4a7966214fd64a731b7a/sql/table.go#L803)):*
You're right!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126)*
<!-- Sent from Reviewable.io -->

kkaneda commented on sql/descriptor.go:
```diff
@@ -136,8 +136,8 @@ func (p *planner) getDescriptor(plainKey descriptorKey, descriptor descriptorPro
 	desc := &Descriptor{}
-	if err := p.txn.GetProto(descKey, desc); err != nil {
-		return err
+	if pErr := p.txn.GetProto(descKey, desc); pErr != nil {
```
My previous commit didn't change `err != nil` to `pErr != nil`... Fixed.

kkaneda commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/executor.go, line 633 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-K9hRxdTY_u6JBzaY9ju:-K9lvEj3sXxF3kplEo83:1599646461) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e3b6ab19b5bc4edd3dc4a7966214fd64a731b7a/sql/executor.go#L633)):*
Oh I see. Interesting. Don't understand the sql code very well, but I can take a look (probably with a separate PR).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/executor.go, line 633 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126#-K9hRxdTY_u6JBzaY9ju:-K9lwLF0B68-kT3ovpjN:677042770) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e3b6ab19b5bc4edd3dc4a7966214fd64a731b7a/sql/executor.go#L633)):*
SGTM

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4126)*
<!-- Sent from Reviewable.io -->
